File: FunctionalProgramming/FunctionalSort.FSharp/FunctionalLibrary/MergeSort.fs
module MergeSort

let rec merge (a: List<'T>) (b: List<'T>) greaterThan =
    if a.IsEmpty then
        b
    else if b.IsEmpty then
        a
    else if greaterThan a.Head b.Head then
        b.Head :: (merge a b.Tail greaterThan)
    else 
        a.Head :: merge a.Tail b greaterThan

let rec sort (list: List<'T>) greaterThan =
    if list |> Seq.length < 2 then
        list
    else 
        let half = (list |> Seq.length) / 2
        merge (sort (list |> List.skip(half)) greaterThan) (sort (list |> List.take(half)) greaterThan) greaterThan


//let rec merge (a: List<string>) (b: List<string>) greaterThan =
//    if a.IsEmpty then
//        b
//    else if b.IsEmpty then
//        a
//    else if greaterThan a.Head  b.Head then
//            b.Head :: (merge a b.Tail greaterThan)
//    else 
//            a.Head :: merge a.Tail b greaterThan
//
//
//let rec sort (list: List<string>) greaterThan =
//    if list |> Seq.length < 2 then
//        list
//    else 
//        let half = (list |> Seq.length) / 2
//        merge (sort (list |> List.skip(half)) greaterThan) (sort (list |> List.take(half)) greaterThan) greaterThan


let rec mergeAlphabetical (a: List<string>) (b: List<string>) =
    if a.IsEmpty then
        b
    else if b.IsEmpty then
        a
    else if a.Head < b.Head then
        a.Head :: mergeAlphabetical a.Tail b
    else 
        b.Head :: mergeAlphabetical a b.Tail

let rec sortAlphabetical (list: List<string>) =
    if list |> Seq.length < 2 then
        list
    else 
        let half = (list |> Seq.length) / 2
        mergeAlphabetical (sortAlphabetical (list |> List.skip(half))) (sortAlphabetical (list |> List.take(half)))



 

File: FunctionalProgramming/FunctionalSort.CSharp/FunctionalLibrary/MergeSort.cs
using System;

namespace FunctionalLibrary
{
    public static class MergeSort
    {
        public static FList<T> Sort<T>(FList<T> list, Func<T, T, bool> greaterThan)
        {
            if (list.Count() < 2)
            {
                return list;
            }
            else
            {
                var half = list.Count() / 2;
                return Merge(Sort(list.Skip(half), greaterThan), Sort(list.Take(half), greaterThan), greaterThan);
            }
        }

        public static FList<T> Merge<T>(FList<T> a, FList<T> b, Func<T, T, bool> greaterThan)
        {
            if (a.IsEmpty)
            {
                return b;
            }
            else if (b.IsEmpty)
            {
                return a;
            }
            else if (greaterThan(a.Head, b.Head))
            {
                return FList.Cons(a.Head, Merge(a.Tail, b, greaterThan));
            }
            else
            {
                return FList.Cons(b.Head, Merge(a, b.Tail, greaterThan));
            }
        }

        public static FList<string> Sort(FList<string> list, Func<string, string, bool> greaterThan)
        {
            if (list.Count() < 2)
            {
                return list;
            }
            else
            {
                var half = list.Count() / 2;
                return Merge(Sort(list.Skip(half), greaterThan), Sort(list.Take(half), greaterThan), greaterThan);
            }
        }

        public static FList<string> Merge(FList<string> a, FList<string> b, Func<string, string, bool> greaterThan)
        {
            if (a.IsEmpty)
            {
                return b;
            }
            else if (b.IsEmpty)
            {
                return a;
            }
            else if (greaterThan(a.Head, b.Head))
            {
                return FList.Cons(a.Head, Merge(a.Tail, b, greaterThan));
            }
            else
            {
                return FList.Cons(b.Head, Merge(a, b.Tail, greaterThan));
            }
        }

        public static FList<string> SortAlphabetical(FList<string> list)
        {
            if (list.Count() < 2)
            {
                return list;
            }
            else
            {
                var half = list.Count() / 2;
                return MergeAlphabetical(SortAlphabetical(list.Skip(half)), SortAlphabetical(list.Take(half)));
            }
        }

        public static FList<string> MergeAlphabetical(FList<string> a, FList<string> b)
        {
            if (a.IsEmpty)
            {
                return b;
            }
            else if (b.IsEmpty)
            {
                return a;
            }
            else if (string.Compare(a.Head, b.Head) < 0)
            {
                return FList.Cons(a.Head, MergeAlphabetical(a.Tail, b));
            }
            else
            {
                return FList.Cons(b.Head, MergeAlphabetical(a, b.Tail));
            }
        }
    }
}
